osdk/src/commands/build/mod.rs

```diff
@@ -3,134 +3,199 @@
+        .clone()
+        .unwrap_or(cargo_target_directory.join(DEFAULT_TARGET_RELPATH));
+    if !osdk_output_directory.exists() {
+        std::fs::create_dir_all(&osdk_output_directory).unwrap();
     }
     let target_info = get_current_crate_info();
-    let bundle_path = osdk_target_directory.join(target_info.name);
+    let bundle_path = osdk_output_directory.join(target_info.name);
+
+    let action = if build_args.for_test {
+        ActionChoice::Test
+    } else {
+        ActionChoice::Run
+    };
 
-    let _bundle = create_base_and_build(
+    let _bundle = create_base_and_cached_build(
         bundle_path,
-        &osdk_target_directory,
-        &ws_target_directory,
+        &osdk_output_directory,
+        &cargo_target_directory,
         config,
+        action,
         &[],
     );
 }
 
-pub fn create_base_and_build(
+pub fn create_base_and_cached_build(
     bundle_path: impl AsRef<Path>,
-    osdk_target_directory: impl AsRef<Path>,
+    osdk_output_directory: impl AsRef<Path>,
     cargo_target_directory: impl AsRef<Path>,
-    config: &BuildConfig,
+    config: &Config,
+    action: ActionChoice,
     rustflags: &[&str],
 ) -> Bundle {
-    let base_crate_path = osdk_target_directory.as_ref().join("base");
+    let base_crate_path = osdk_output_directory.as_ref().join("base");
     new_base_crate(
         &base_crate_path,
         &get_current_crate_info().name,
         get_current_crate_info().path,
     );
     let original_dir = std::env::current_dir().unwrap();
     std::env::set_current_dir(&base_crate_path).unwrap();
-    let bundle = do_build(
+    let bundle = do_cached_build(
         &bundle_path,
-        &osdk_target_directory,
+        &osdk_output_directory,
         &cargo_target_directory,
         config,
+        action,
         rustflags,
     );
     std::env::set_current_dir(original_dir).unwrap();
     bundle
 }
 
+/// If the source is not since modified and the last build is recent, we can reuse the existing bundle.
+pub fn do_cached_build(
+    bundle_path: impl AsRef<Path>,
+    osdk_output_directory: impl AsRef<Path>,
+    cargo_target_directory: impl AsRef<Path>,
+    config: &Config,
+    action: ActionChoice,
+    rustflags: &[&str],
+) -> Bundle {
+    let build_a_new_one = || {
+        do_build(
+            &bundle_path,
+            &osdk_output_directory,
+            &cargo_target_directory,
+            config,
+            action,
+            rustflags,
+        )
+    };
+
+    let existing_bundle = Bundle::load(&bundle_path);
+    let Some(existing_bundle) = existing_bundle else {
+        return build_a_new_one();
+    };
+    if existing_bundle.can_run_with_config(config, action).is_err() {
+        return build_a_new_one();
+    }
+    let Ok(built_since) = SystemTime::now().duration_since(existing_bundle.last_modified_time())
+    else {
+        return build_a_new_one();
+    };
+    if built_since > Duration::from_secs(600) {
+        return build_a_new_one();
+    }
```

Comment:
Why do we need to set a timeout here? If the code is not modified, we can always reuse the cached results, right?

Reply:
No, building somewhat depends on "something in the cloud" since cargo will fetch the git dependencies even if cached.
Git dependencies are not properly locked. In the future we may move to `crates.io` as more as possible.